src/Raven.Server/Documents/ShardedHandlers/ShardedBatchHandler.cs

```diff
@@ -0,0 +1,355 @@
+            var commandBuilder = new ShardedBatchCommandBuilder(this, ShardedContext.DatabaseName, ShardedContext.IdentitySeparator);
+            using (ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            {
+                var contentType = HttpContext.Request.ContentType;
+                if (contentType == null ||
+                    contentType.StartsWith("application/json", StringComparison.OrdinalIgnoreCase))
+                {
+                    await commandBuilder.BuildCommandsAsync(context, RequestBodyStream());
+                }
+                else if (contentType.StartsWith("multipart/mixed", StringComparison.OrdinalIgnoreCase) ||
+                         contentType.StartsWith("multipart/form-data", StringComparison.OrdinalIgnoreCase))
+                {
+                    await commandBuilder.ParseMultipart(context, RequestBodyStream(), HttpContext.Request.ContentType);
+                }
+                else
+                    BatchHandler.ThrowNotSupportedType(contentType);
+
+                if (TrafficWatchManager.HasRegisteredClients)
+                {
+                    var log = BatchHandler.BatchTrafficWatch(commandBuilder.Commands);
+                    // add sb to httpContext
+                    AddStringToHttpContext(log, TrafficWatchChangeType.BulkDocs); // TODO sharding: maybe ShardedBulkDocs type?
+                }
+
+                using (var batch = await commandBuilder.GetCommand(context))
+                {
+                    if (batch.IsClusterTransaction)
+                    {
+                        throw new NotSupportedException("Cluster transactions are currently not supported in a sharded database.");
+                    }
+
+                    var shardedBatchCommands = new Dictionary<int, SingleNodeShardedBatchCommand>();
+                    foreach (var command in batch)
+                    {
+                        var id = command.Id;
+                        var shardIndex = ShardedContext.GetShardIndex(context, id);
+                        var requestExecutor = ShardedContext.RequestExecutors[shardIndex];
+
+                        if (shardedBatchCommands.ContainsKey(shardIndex) == false)
+                        {
+                            shardedBatchCommands.Add(shardIndex, new SingleNodeShardedBatchCommand(this, requestExecutor.ContextPool));
+                        }
+
+                        var shardedBatchCommand = shardedBatchCommands[shardIndex];
+                        shardedBatchCommand.AddCommand(command);
+                    }
+
+                    var tasks = shardedBatchCommands.Select(x => ShardedContext.RequestExecutors[x.Key].ExecuteAsync(x.Value, x.Value.Context));
+                    await Task.WhenAll(tasks);
+
+                    HttpContext.Response.StatusCode = (int)HttpStatusCode.Created;
+                    using (var writer = new BlittableJsonTextWriter(context, ResponseBodyStream()))
```

Comment:
How do we ensure that we get the results in the same order that the user asked for them?

Reply:
This is actually a good point. I know it is a requirement for `Load` (need to check if we do that properly there as well)
But is that a requirement for bulk docs as well?

Reply:
It is a *hard* requirement for bulk docs as well.
On the client side, we are assuming that the order of results match the order of commands.

Reply:
done by marking for each request the position in the response